<launch>

    <arg name="manual_control" default="false" />

    <!-- Launch the simulator -->
    <include file="$(find caric_mission)/launch/run_mbs.launch" />
    <!-- <include file="$(find caric_mission)/launch/run_hangar.launch"/> -->
    <!-- <include file="$(find caric_mission)/launch/run_crane.launch"/> -->

    <!-- Central coordinator -->
    <node name="gcs_task" pkg="caric_baseline" output="screen" type="gcs_task"  />
    <!-- <node name="gcs_talker" pkg="caric_baseline" type="initial_task"  /> -->

    <group ns="jurong">
        
        <arg name="namespace"  default="jurong" />

        <node name="baseline_planner" pkg="caric_baseline" type="baseline_planner" launch-prefix="bash -c 'sleep 5; $0 $@'" >

            <!-- Input topics -->
            <remap from="/cloud_inW"                   to="/$(arg namespace)/cloud_inW" />
            <remap from="/nbr_odom_cloud"              to="/$(arg namespace)/nbr_odom_cloud" />
            <remap from="/firefly/gimbal"              to="/$(arg namespace)/gimbal" />
            <remap from="/ground_truth/odometry"       to="/$(arg namespace)/ground_truth/odometry" />

            <!-- Control outputs -->
            <remap from="/firefly/command/gimbal"      to="/$(arg namespace)/command/gimbal" />
            <remap from="/firefly/command/trajectory"  to="/$(arg namespace)/command/trajectory" />

            <!-- Vizualization -->
            <remap from="/firefly/map"                 to="/$(arg namespace)/map" />
            <remap from="/firefly/Nbr_mask"            to="/$(arg namespace)/Nbr_mask" />
            <remap from="/firefly/path_show"           to="/$(arg namespace)/path_show" />
        </node>

    </group>

    <group ns="changi">
    
        <arg name="namespace"  default="changi" />

        <node name="baseline_planner" pkg="caric_baseline" type="baseline_planner" launch-prefix="bash -c 'sleep 5; $0 $@'" >

            <!-- Input topics -->
            <remap from="/cloud_inW"                   to="/$(arg namespace)/cloud_inW" />
            <remap from="/nbr_odom_cloud"              to="/$(arg namespace)/nbr_odom_cloud" />
            <remap from="/firefly/gimbal"              to="/$(arg namespace)/gimbal" />
            <remap from="/ground_truth/odometry"       to="/$(arg namespace)/ground_truth/odometry" />

            <!-- Control outputs -->
            <remap from="/firefly/command/gimbal"      to="/$(arg namespace)/command/gimbal" />
            <remap from="/firefly/command/trajectory"  to="/$(arg namespace)/command/trajectory" />

            <!-- Vizualization -->
            <remap from="/firefly/map"                 to="/$(arg namespace)/map" />
            <remap from="/firefly/Nbr_mask"            to="/$(arg namespace)/Nbr_mask" />
            <remap from="/firefly/path_show"           to="/$(arg namespace)/path_show" />
        </node>

    </group>
    
    <group ns="raffles">
        
        <arg name="namespace"  default="raffles" />

        <node name="baseline_planner" pkg="caric_baseline" type="baseline_planner" launch-prefix="bash -c 'sleep 5; $0 $@'" output="screen" >
            
            <!-- Input topics -->
            <remap from="/cloud_inW"                   to="/$(arg namespace)/cloud_inW" />
            <remap from="/nbr_odom_cloud"              to="/$(arg namespace)/nbr_odom_cloud" />
            <remap from="/firefly/gimbal"              to="/$(arg namespace)/gimbal" />
            <remap from="/ground_truth/odometry"       to="/$(arg namespace)/ground_truth/odometry" />

            <!-- Control outputs -->
            <remap from="/firefly/command/gimbal"      to="/$(arg namespace)/command/gimbal" />
            <remap from="/firefly/command/trajectory"  to="/$(arg namespace)/command/trajectory" />

            <!-- Vizualization -->
            <remap from="/firefly/map"                 to="/$(arg namespace)/map" />
            <remap from="/firefly/Nbr_mask"            to="/$(arg namespace)/Nbr_mask" />
            <remap from="/firefly/path_show"           to="/$(arg namespace)/path_show" />
        </node>

    </group>

    <group ns="sentosa">
        
        <arg name="namespace"  default="sentosa" />

        <node name="baseline_planner" pkg="caric_baseline" type="baseline_planner" launch-prefix="bash -c 'sleep 5; $0 $@'" >

            <!-- Input topics -->
            <remap from="/cloud_inW"                   to="/$(arg namespace)/cloud_inW" />
            <remap from="/nbr_odom_cloud"              to="/$(arg namespace)/nbr_odom_cloud" />
            <remap from="/firefly/gimbal"              to="/$(arg namespace)/gimbal" />
            <remap from="/ground_truth/odometry"       to="/$(arg namespace)/ground_truth/odometry" />

            <!-- Control outputs -->
            <remap from="/firefly/command/gimbal"      to="/$(arg namespace)/command/gimbal" />
            <remap from="/firefly/command/trajectory"  to="/$(arg namespace)/command/trajectory" />

            <!-- Vizualization -->
            <remap from="/firefly/map"                 to="/$(arg namespace)/map" />
            <remap from="/firefly/Nbr_mask"            to="/$(arg namespace)/Nbr_mask" />
            <remap from="/firefly/path_show"           to="/$(arg namespace)/path_show" />
        </node>

    </group>
    
    <group ns="nanyang">
        
        <arg name="namespace"  default="nanyang" />

        <node name="baseline_planner" pkg="caric_baseline" type="baseline_planner" launch-prefix="bash -c 'sleep 5; $0 $@'" >

            <!-- Input topics -->
            <remap from="/cloud_inW"                   to="/$(arg namespace)/cloud_inW" />
            <remap from="/nbr_odom_cloud"              to="/$(arg namespace)/nbr_odom_cloud" />
            <remap from="/firefly/gimbal"              to="/$(arg namespace)/gimbal" />
            <remap from="/ground_truth/odometry"       to="/$(arg namespace)/ground_truth/odometry" />

            <!-- Control outputs -->
            <remap from="/firefly/command/gimbal"      to="/$(arg namespace)/command/gimbal" />
            <remap from="/firefly/command/trajectory"  to="/$(arg namespace)/command/trajectory" />

            <!-- Vizualization -->
            <remap from="/firefly/map"                 to="/$(arg namespace)/map" />
            <remap from="/firefly/Nbr_mask"            to="/$(arg namespace)/Nbr_mask" />
            <remap from="/firefly/path_show"           to="/$(arg namespace)/path_show" />
        </node>

    </group>
    
    <!-- For visualization -->
    <node name="odom2tf" pkg="caric_mission" type="odom2tf.py"  />

</launch>